### PR TITLE
Allow work with VPCs on user-mode (issue #81)

### LINKED ISSUE #81: Allow work with VPCs on user-mode
# Expected Behavior

# Current Behavior

# Possible Solution

# Steps to Reproduce (for bugs)

    1.
    2.
    3.
    4.

# Context and environment

### PR DESCRIPTION
**Pull request by pcantera**
Closes #81

imagen

imagen

imagen

# Description

# Related Issue

# Motivation and Context

# How Has This Been Tested?

# Screenshots (if appropriate)

# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by chris-m-h — APPROVED** [review]

**Review by pbanos — CHANGES_REQUESTED** [review]
Looks good, but it is not any use to manage vpcs if you cannot manage subnets, I think it should be part of the same issue

**Comment by pcantera** [comment]
> Looks good, but it is not any use to manage vpcs if you cannot manage subnets, I think it should be part of the same issue

Yes, _there's another opened issue about subnets_.  This was done in a time window to go ahead with these functionalities.The idea is to include Subnet asap.

Do you prefer to integrate it in current issue?
[EDIT] No subnet issue really! but i have it in my roadmap. so i'm going to integrate it here asap

**Comment by pcantera** [comment]
@pbanos Added Subnet management

**Review by sjpatino — APPROVED** [review]
:+1:

**Review by pbanos — CHANGES_REQUESTED** [review]
As we talked last Friday, we need to include VPN management. Also, missing from both documentation and actual routes, we need a /network/subnets/:subnet_id/servers resource that lists the servers inside a subnet: once that resource is done it should be integrated in this PR as well.

**Comment by pcantera** [comment]
@pbanos Added:
- VPC discard method
- Subnet list-servers method
- VPN management

File: api/network/vpcs_api.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	if err = utils.CheckStandardStatus(status, data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DiscardVPC deletes a VPC by its ID
```
Comment (pcantera) [inline api/network/vpcs_api.go]: detected... should be 'discards'

**Comment by lttito** [comment]
please @pcantera remember to merge PR https://github.com/ingrammicro/tapp/pull/1524 at the same time as this PR.
API documentation is updated

**Review by lttito — APPROVED** [review]
:+1:

File: api/network/vpns_api.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	if err = utils.CheckStandardStatus(status, data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetVPNListPlans returns the list of VPN plans for a given VPC ID
+func (dm *VPNService) GetVPNListPlans(vpcID string) (vpnPlans []*types.VpnPlan, err error) {
```
Comment (pbanos) [inline api/network/vpns_api.go]: This is not good naming, it should be GetVPNPlanList. As a future refactor (in a separate issue) I propose renaming all these `Get<resource>List` methods as `List<resource>s` though.

File: api/types/subnets.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+// Subnet stores an IMCO VPC Subnet item
+type Subnet struct {
+	ID                     string `json:"id" header:"ID"`
+	Name                   string `json:"name,omitempty" header:"NAME"`
+	CIDR                   string `json:"cidr,omitempty" header:"CIDR"`
+	State                  string `json:"state,omitempty" header:"STATE"`
+	Type                   string `json:"type,omitempty" header:"STATE"`
+	VpcID                  string `json:"vpc_id,omitempty" header:"VPC_ID"`
+	ServerCreationDisabled bool   `json:"server_creation_disabled,omitempty" header:"SERVER_CREATION_DISABLED" show:"nolist,noshow"`
```
Comment (pbanos) [inline api/types/subnets.go:11]: I'd change the header to `READY FOR SERVERS`

**Review by pbanos — CHANGES_REQUESTED** [review]
Small change requests. I think that other than that it is ready to be merged. Good work! :+1:

File: api/network/vpns_api.go (hunk 2 above)
Comment (pcantera) [reply api/network/vpns_api.go]: Done

File: api/types/subnets.go (hunk 3 above)
Comment (pcantera) [reply api/types/subnets.go:11]: As we have commented, left it as is, since this field is not really shown in user-mode.

**Review by pbanos — APPROVED** [review]
:+1:

**Pull request merged**